Checks/Compose/PatternLengthCheck.cs

```diff
@@ -0,0 +1,219 @@
+                {
+                    foreach (var snapValues in shortSnapParams[diff])
+                    {
+                        var currentPatternStartTimeMs = objects.FirstOrDefault()?.time ?? 0;
+                        var currentPatternEndTimeMs = objects.FirstOrDefault()?.time ?? 0;
+
+                        // variables to identify length of pattern
+                        var patternStartIndex = 0;
+                        var foundStartOfPattern = false;
+                        var foundEndOfPattern = false;
+                        for (int i = 0; i < objects.Count; i++)
+                        {
+                            var current = objects.SafeGetIndex(i);
+                            var timing = beatmap.GetTimingLine<UninheritedLine>(current.time);
+                            var normalizedMsPerBeat = timing.GetNormalizedMsPerBeat();
+
+                            // convert minimal gap beats to milliseconds
+                            var snapMs = snapValues.Key * normalizedMsPerBeat;
+
+
+                            // check if this is end of pattern
+                            if (i + 1 < objects.Count && foundStartOfPattern)
+                            {
+                                var gapBeginObject = objects.SafeGetIndex(i);
+                                var gapEndObject = objects.SafeGetIndex(i + 1);
+
+                                // Check if gap is greater than the snap size
+                                var gap = gapEndObject.time - gapBeginObject.GetEndTime();
+                                if (gap - MS_EPSILON > snapMs)
+                                {
+                                    foundEndOfPattern = true;
+                                    currentPatternEndTimeMs = gapBeginObject.GetEndTime();
+                                }
+                            } else if (i == objects.Count - 1 && foundStartOfPattern)
+                            {
+                                // last note, so forced end of pattern
+                                foundEndOfPattern = true;
+                                currentPatternEndTimeMs = objects.SafeGetIndex(i).GetEndTime();
+                            }
+
+                            // check if this is start of pattern
+                            if (i + 1 < objects.Count && !foundStartOfPattern)
+                            {
+                                var gapBeginObject = objects.SafeGetIndex(i);
+                                var gapEndObject = objects.SafeGetIndex(i + 1);
+
+                                // Check if gap is smaller than or equal to the snap size
+                                var gap = gapEndObject.time - gapBeginObject.GetEndTime();
+                                if (gap - MS_EPSILON <= snapMs)
+                                {
+                                    foundStartOfPattern = true;
+                                    currentPatternStartTimeMs = gapBeginObject.GetEndTime();
```

Comment:
nitpick: Technically more correct to use 'gapBeginObject.time' here instead of end time, since this is the start of the pattern.

Doesn't actually make a difference though since you're already filtering the objects to circles only but it would be less fragile if you fixed this in case for some weird reason we ever included non-circles in this check

Reply:
Makes sense!